### PR TITLE
92 replace exchange rate and current tokenworth with current gains

### PR DESCRIPTION
**Pull request by aliberkhsl**
#92

### CODE REVIEW

File: contracts/staking/GoodCompoundStaking.sol [hunk 1]
```diff
@@ -141,6 +133,51 @@ contract GoodCompoundStaking is SimpleStaking {
...
+		} else {
+			tokenBalance =
+				((iToken.balanceOf(address(this)) / (10**decimalDifference)) *
+					er) /
+				10**mantissa; // based on https://compound.finance/docs#protocol-math
+		}
+		if (tokenBalance <= totalProductivity) {
+			return (0, 0, tokenBalance);
+		}
+		uint256 tokenGains = tokenBalance - (totalProductivity);
```
Comment (sirpy) [inline contracts/staking/GoodCompoundStaking.sol:166]: remove extra (

File: contracts/staking/GoodCompoundStaking.sol [hunk 2]
```diff
@@ -149,4 +186,26 @@ contract GoodCompoundStaking is SimpleStaking {
 	{
 		return collectInterestGasCost;
 	}
+
+	/**
+	 * @dev Calculates worth of given amount of iToken in Token
+	 * @param _amount Amount of token to calculate worth in Token
+	 * @return Worth of given amount of token in Token
+	 */
+	function iTokenWorthinToken(uint256 _amount)
```
Comment (sirpy) [inline contracts/staking/GoodCompoundStaking.sol:195]:
```suggestion
	function iTokenWorthInToken(uint256 _amount)
```

File: contracts/staking/GoodCompoundStaking.sol [hunk 3]
```diff
@@ -141,6 +133,51 @@ contract GoodCompoundStaking is SimpleStaking {
...
+				(iToken.balanceOf(address(this)) *
+					(10**decimalDifference) *
+					er) /
+				10**mantissa; // based on https://compound.finance/docs#protocol-math
+		} else {
+			tokenBalance =
+				((iToken.balanceOf(address(this)) / (10**decimalDifference)) *
+					er) /
+				10**mantissa; // based on https://compound.finance/docs#protocol-math
+		}
```
Comment (sirpy) [inline contracts/staking/GoodCompoundStaking.sol:162]: DRY this seems like iTokenWorthInToken

File: contracts/staking/SimpleStaking.sol [hunk 4]
```diff
@@ -377,28 +429,10 @@ contract SimpleStaking is
...
-				((tokenGains / 10**decimalDifference) * 10**mantissa) /
-				er; // based on https://compound.finance/docs#protocol-math
-		} else {
-			iTokenGains =
-				((tokenGains * 10**decimalDifference) * 10**mantissa) /
-				er; // based on https://compound.finance/docs#protocol-math
-		}
+		(uint256 iTokenGains, uint256 tokenGains, ) = currentGains();
 		tokenGains = getTokenValueInUSD(tokenGains);
 		return (iTokenGains, tokenGains);
```
Comment (sirpy) [inline contracts/staking/SimpleStaking.sol:437]: should return triplet
itokengains,tokengains,usdgains

see also comment on interestcollected event

File: contracts/staking/SimpleStaking.sol [hunk 5]
```diff
@@ -19,12 +18,7 @@ import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
  * or withdraw their stake in Tokens
  * the contracts buy intrest tokens and can transfer the daily interest to the  DAO
  */
-contract SimpleStaking is
-	AbstractGoodStaking,
-	ERC20Upgradeable,
-	DAOContract,
-	BaseShareField
-{
+contract SimpleStaking is ERC20Upgradeable, DAOContract, BaseShareField {
```
Comment (sirpy) [inline contracts/staking/SimpleStaking.sol:21]: make abstract
and dont use empty implementation for all unimplemented methods

File: contracts/staking/SimpleStaking.sol [hunk 6]
```diff
@@ -46,6 +40,33 @@ contract SimpleStaking is
...
+	/**
+	 * @dev Emitted when fundmanager transfers intrest collected from defi protrocol.
+	 * `recipient` will receive `intrestTokenValue` as intrest.
+	 */
+	event InterestCollected(
+		address recipient,
+		address token,
+		address intrestToken,
+		uint256 intrestTokenValue,
+		uint256 tokenValue
```
Comment (sirpy) [inline contracts/staking/SimpleStaking.sol:68]: emit just recipient, itokenAmount,tokenAmount,usdValue

File: contracts/staking/SimpleStaking.sol [hunk 7]
```diff
@@ -46,6 +40,33 @@ contract SimpleStaking is
...
+
+	/**
+	 * @dev Emitted when `staker` withdraws their stake `value` tokens and contracts balance will
+	 * be reduced to`remainingBalance`.
+	 */
+	event StakeWithdraw(
+		address indexed staker,
+		address token,
+		uint256 value,
+		uint256 remainingBalance
```
Comment (sirpy) [inline contracts/staking/SimpleStaking.sol:56]: remove remainingbalance

**Review by sirpy — CHANGES_REQUESTED** [review]
remove end() it just calls pause, add a comment for pause

File: contracts/staking/GoodCompoundStaking.sol (hunk 2 above)
Comment (aliberkhsl) [reply contracts/staking/GoodCompoundStaking.sol:195]: replaced

File: contracts/staking/GoodCompoundStaking.sol (hunk 1 above)
Comment (aliberkhsl) [reply contracts/staking/GoodCompoundStaking.sol:166]: removed

File: contracts/staking/GoodCompoundStaking.sol (hunk 3 above)
Comment (aliberkhsl) [reply contracts/staking/GoodCompoundStaking.sol:162]: removed

File: contracts/staking/SimpleStaking.sol (hunk 6 above)
Comment (aliberkhsl) [reply contracts/staking/SimpleStaking.sol:68]: changed

File: contracts/staking/SimpleStaking.sol (hunk 7 above)
Comment (aliberkhsl) [reply contracts/staking/SimpleStaking.sol:56]: removed

File: contracts/staking/SimpleStaking.sol (hunk 4 above)
Comment (aliberkhsl) [reply contracts/staking/SimpleStaking.sol:437]: changed it

File: contracts/staking/SimpleStaking.sol [hunk 8]
```diff
@@ -424,17 +417,21 @@ contract SimpleStaking is ERC20Upgradeable, DAOContract, BaseShareField {
 
 	/**
 	 * @dev Calculates the current interest that was gained.
-	 * @return (uint256, uint256, uint256) The interest in iToken, the interest in USD
+	 * @return (uint256, uint256, uint256) The interest in iToken, the interest in Token and the interest in USD
 	 */
 	function currentUBIInterest()
```
Comment (sirpy) [inline contracts/staking/SimpleStaking.sol:419]: this seems  like it can be merged to currentGains

**Review by sirpy — CHANGES_REQUESTED** [review]

File: contracts/staking/SimpleStaking.sol (hunk 8 above)
Comment (aliberkhsl) [reply contracts/staking/SimpleStaking.sol:419]: Removed and integrated it in currentGains

**Review by sirpy — CHANGES_REQUESTED** [review]
i dont see a test for true,true - false ,false

**Comment by aliberkhsl** [comment]
> i dont see a test for true,true - false ,false

Added

**Pull request merged**